library/src/main/scala/scala/scalajs/macroImpls/UseAsMacros.scala

```diff
@@ -0,0 +1,304 @@
+      c: Context { type PrefixType = js.Using[_] }): c.Expr[B] = {
+    (new Macros[c.type](c)).as[A, B]
+  }
+
+  private class Macros[C <: Context { type PrefixType = js.Using[_] }](val c: C)
+      extends JSMembers with Compat210Component {
+
+    import c.universe._
+
+    private val JSNameAnnotation = typeOf[js.annotation.JSName].typeSymbol
+    private val JSExportAnnotation = typeOf[js.annotation.JSExport].typeSymbol
+    private val JSExportAllAnnotation = typeOf[js.annotation.JSExportAll].typeSymbol
+
+    /** Base classes that are allowed in a target type.
+     *  These are also the classes whose methods do not need to be provided.
+     */
+    private val JSObjectAncestors = typeOf[js.Object].baseClasses.toSet
+
+    def as[A : WeakTypeTag, B <: js.Any : WeakTypeTag]: Expr[B] = {
+      val trgTpe = verifyTargetType(weakTypeOf[B])
+      val srcTpe = weakTypeOf[A]
+
+      check(srcTpe, trgTpe)
+
+      reify { c.prefix.splice.x.asInstanceOf[B] }
+    }
+
+    /** Perform the actual structural typechecking.
+     *
+     *  Checks if [[srcTpe]] conforms to [[trgTpe]]. Reports errors otherwise.
+     */
+    private def check(srcTpe: Type, trgTpe: Type): Unit = {
+      val rawMembers = rawJSMembers(trgTpe)
+
+      /* Nothing exports everything, but we still need to calculate the members
+       * of the raw JS class, since they might be malformed
+       * (i.e. apply or bad refinements)
+       */
+      if (srcTpe.typeSymbol != definitions.NothingClass) {
+        val exports = exportedMembers(srcTpe)
+        checkMembers(srcTpe, exports, rawMembers)
+      }
+    }
+
+    /** Checks that the [[exports]] conform to the [[rawMembers]].
+     *  [[srcTpe]] is only used for error messages.
+     */
+    private def checkMembers(srcTpe: Type, exports: Map[String, List[JSMember]],
+        rawMembers: Iterable[(String, JSMember)]): Unit = {
+
+      for ((name, jsMem) <- rawMembers) {
+        exports.get(name).fold {
+          val msg = s"$srcTpe does not export a member named $name"
+          c.error(c.enclosingPosition, msg)
+        } { candidates =>
+          if (!hasConformingMembers(jsMem, candidates)) {
+            val msg = s"$srcTpe does not export " + jsMem.displayStr(name)
+            c.error(c.enclosingPosition, msg)
+          }
+        }
+      }
+    }
+
+    /** Whether the [[candidates]] together conform to [[jsMem]] */
+    private def hasConformingMembers(jsMem: JSMember,
+        candidates: List[JSMember]): Boolean = {
+
+      def find(mem: JSMember) = candidates.exists(_.conformsTo(mem))
+
+      def hasConformingOverloads = jsMem match {
+        case jsMem @ JSMethod(params, _) if params.last.isDefault =>
+          /* If the last parameter is a default parameter, the method can also
+           * be implemented by two methods, one with and one without the
+           * parameter. Note that this does not generalize to more parameters
+           * due to named parameters.
```

Comment:
Yes, I think we should drop it. Keep it safe in a commit somewhere, in case we need it later.